lib/luaxtable.c

```diff
@@ -32,48 +32,127 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
-{
-	return true;
-}
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+} luaxtable_hooks = {NULL, NULL};
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const char *op, int nret)
 {
-	return 0;
-}
+	int ret = -1;
+	int nargs = 0;
+	lunatik_object_t *data = NULL;
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
-{
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	if (skb != NULL ){
+		if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA) {
+			pr_err("%s: could not find skb\n", op);
+			goto err;
+		}
+
+		data = (lunatik_object_t *)lunatik_toobject(L, -1);
+		if (skb_is_nonlinear(skb))
+			pr_err("fragemented skb found!\n");
+
+		luadata_reset(data, skb->data, skb_headlen(skb));
+		nargs = 1;
+	}
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+	if (nret != 0)
+		ret = luaL_optinteger(L, -1, 0);
+
+err:
+	if (data)
+		luadata_clear(data);
+	return ret;
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
-{
-	return NF_ACCEPT;
+static bool match_fallback = false;
+static unsigned int target_fallback = XT_CONTINUE;
```

Comment:
as already pointed out, you should prefix every global with `luaxtable_`.. in this case, I think it's a good idea to follow the same pattern you have done for the RCU tables, using a struct.. perhaps we can have a single struct to hold table pointe and the default return.. it's up to you